pkg/service/auth_context.go

```diff
@@ -47,71 +48,122 @@ func NewAuthContext(parentCtx context.Context, req *auth.CheckRequest, apiConfig
 
 }
 
-func (authContext *AuthContext) evaluateAuthConfig(ctx context.Context, config common.AuthConfigEvaluator, cb configCallback) error {
+func (authContext *AuthContext) evaluateAuthConfig(ctx context.Context, config common.AuthConfigEvaluator, cb evaluateCallback) error {
 	select {
 	case <-ctx.Done():
-		authCxtLog.Info("Context cancelled objConfig terminating", "config", config)
+		authCtxLog.Info("Context aborted", "config", config)
 		return nil
 	default:
-		if authObj, err := config.Call(authContext); err != nil {
-			authCxtLog.Error(err, "Invalid auth object config")
-			return err
-		} else {
-			cb(config, authObj)
-			return nil
-		}
+		authObj, err := config.Call(authContext)
+		return cb(config, authObj, err)
 	}
 }
 
-func (authContext *AuthContext) evaluateAuthConfigs(authConfigs []common.AuthConfigEvaluator, cb configCallback) error {
-	errGroup, ctx := errgroup.WithContext(context.Background())
+func (authContext *AuthContext) evaluateAuthConfigs(ctx context.Context, authConfigs []common.AuthConfigEvaluator, cb evaluateCallback) error {
+	errGroup, errCtx := errgroup.WithContext(ctx)
 
 	for _, authConfig := range authConfigs {
 		objConfig := authConfig
 		errGroup.Go(func() error {
-			return authContext.evaluateAuthConfig(ctx, objConfig, cb)
+			return authContext.evaluateAuthConfig(errCtx, objConfig, cb)
 		})
 	}
 	if err := errGroup.Wait(); err != nil {
 		return err
 	} else {
-		authCxtLog.Info("Successfully fetched all auth objects.", "authConfigs", authConfigs)
+		authCtxLog.Info("Successfully evaluated all auth objects.", "authConfigs", authConfigs)
 		return nil
 	}
 }
 
-// Evaluate evaluates all steps of the auth pipeline (identity → metadata → policy enforcement)
-func (authContext *AuthContext) Evaluate() error {
-	// identity
-	if err := authContext.evaluateAuthConfigs(authContext.API.IdentityConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
-			// Convert from interfaceType to SpecificType
-			v, _ := conf.(*config.IdentityConfig)
-			authCxtLog.Info("Identity", "Config", conf, "AuthObj", authObj)
-			authContext.Identity[v] = authObj
-		}); err != nil {
-		return err
+// EvaluateIdentity evaluates given identity configs and retrieves the first valid identity object.
+func (authContext *AuthContext) EvaluateIdentity() error {
+	identityConfigs := authContext.API.IdentityConfigs
+	if len(identityConfigs) <= 0 {
+		return errors.New("no identity configs found")
 	}
 
-	// metadata
-	if err := authContext.evaluateAuthConfigs(authContext.API.MetadataConfigs,
-		func(conf common.AuthConfigEvaluator, authObj interface{}) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+
+	successChannel := make(chan bool)
+
+	go func() {
+		defer close(successChannel)
```

Comment:
this is very risky. You could be closing some channel and after that try to read it in the main goroutine